File: adsa/Day27(graphinC++)/directedcycle.py
class Solution:
    def dfs(self, node, visited, pathVisited, adj_list):
        visited[node] = True
        pathVisited[node] = True

        for adjNode in adj_list[node]:
            if not visited[adjNode]:
                if self.dfs(adjNode, visited, pathVisited, adj_list):
                    return True
            elif pathVisited[adjNode]:  # Found a back edge
                return True

        pathVisited[node] = False  # Backtrack
        return False

    def isCycle(self, V, edges):
        adj_list = [[] for _ in range(V)]
        for u, v in edges:
            adj_list[u].append(v)  # directed edge

        visited = [False] * V
        pathVisited = [False] * V

        for i in range(V):
            if not visited[i]:
                if self.dfs(i, visited, pathVisited, adj_list):
                    return True
        return False
# Example usage:
edges = [(0, 1), (1, 2), (2, 0)]  # Example directed graph with a cycle
solution = Solution()
print(solution.isCycle(3, edges))  # Output: True

File: adsa/Day27(graphinC++)/directedcycle.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;
class Solution {
  private:
    bool dfscheck(int node,vector<vector<int>> &adj,int vis[],int pathVis[]){
        vis[node]=1;
        pathVis[node]=1;
        for(auto it :adj[node]){
            if(!vis[it]){
                if(dfscheck(it,adj,vis,pathVis)==true) return true;
            }
            else if(pathVis[it]) return true;
            
        }
        pathVis[node]=0;
        return false;
    }
  public:
    bool isCyclic(int V, vector<vector<int>> &edges) {
        // code here
        vector<vector<int>> adj(V);
        int n= edges.size();
        for(int i=0;i<n;i++){
            int u=edges[i][0];
            int v=edges[i][1];
            //adj[v].push_back(u);
            adj[u].push_back(v);
        }
        int vis[V]={0};
        int pathVis[V]={0};
        for(int i=0;i<V;i++){
            if(!vis[i]){
                if(dfscheck(i,adj,vis,pathVis)==true){
                    return true;
                }
            }
        }
        return false;
    }
};

int main() {
    // Example usage
    Solution sol;
    vector<vector<int>> edges = {{0, 1}, {1, 2}, {2, 0}}; // Example directed graph with a cycle
    int V = 3; // Number of vertices
    bool result = sol.isCyclic(V, edges);
    
    // Output the result
    cout << (result ) << endl;
    return 0;
}